File: scripts/sources/ccfddl.py
"""ccfddl/ccf-deadlines source."""

from __future__ import annotations

from pathlib import Path

import yaml

from ..model import Conference, Deadline, Edition, parse_date_range, parse_instant, slug, warn
from .base import fetch_tarball

REPO = "ccfddl/ccf-deadlines"
REF = "main"
NAME = "ccfddl"

# 'abstract deadline' (with a space) exists once upstream.
_ABSTRACT_KEYS = ("abstract_deadline", "abstract deadline")


def _deadlines_of(timeline: list, tz_raw: str) -> list[Deadline]:
    out: list[Deadline] = []
    for index, entry in enumerate(timeline or []):
        if not isinstance(entry, dict):
            continue
        rnd = index + 1
        comment = entry.get("comment")
        raw_abstract = None
        for key in _ABSTRACT_KEYS:
            if entry.get(key) is not None:
                raw_abstract = entry[key]
                break
        for kind, label, raw in (
            ("abstract", "Abstract submission", raw_abstract),
            ("paper", "Paper submission", entry.get("deadline")),
        ):
            if raw is None:
                continue
            at = parse_instant(raw, tz_raw)
            if at is None:
                continue
            out.append(
                Deadline(
                    kind=kind,
                    label=label,
                    at_utc=at,
                    tz_raw=tz_raw,
                    round=rnd,
                    comment=comment,
                )
            )
    return out


def _edition_of(raw: dict) -> Edition | None:
    try:
        year = int(raw["year"])
    except (KeyError, TypeError, ValueError):
        warn(f"ccfddl edition without a usable year: {raw.get('id')!r}")
        return None
    tz_raw = str(raw.get("timezone") or "")
    date_text = str(raw.get("date") or "")
    start, end = parse_date_range(date_text, year)
    return Edition(
        year=year,
        edition_id=str(raw.get("id") or ""),
        link=str(raw.get("link") or ""),
        place=str(raw.get("place") or ""),
        date_text=date_text,
        event_start=start,
        event_end=end,
        deadlines=_deadlines_of(raw.get("timeline"), tz_raw),
        source=NAME,
    )


def _conference_of(raw: dict) -> Conference | None:
    title = str(raw.get("title") or "").strip()
    if not title:
        return None
    editions = [
        e for e in (_edition_of(c) for c in (raw.get("confs") or [])) if e
    ]
    editions.sort(key=lambda e: e.year)
    rank = {
        str(k).lower(): str(v)
        for k, v in (raw.get("rank") or {}).items()
        if v is not None
    }
    link = ""
    for edition in reversed(editions):
        if edition.link:
            link = edition.link
            break
    return Conference(
        key=slug(title),
        title=title,
        full_name=str(raw.get("description") or title),
        link=link,
        rank=rank,
        dblp=raw.get("dblp"),
        upstream_sub=raw.get("sub"),
        editions=editions,
        sources=[NAME],
    )


def parse_tree(conference_dir: Path) -> list[Conference]:
    """Read every ``conference/<SUB>/<name>.yml`` under an extracted tree."""
    out: list[Conference] = []
    for path in sorted(Path(conference_dir).rglob("*.yml")):
        if path.name == "types.yml":
            continue
        try:
            loaded = yaml.safe_load(path.read_text(encoding="utf-8"))
        except yaml.YAMLError as exc:
            warn(f"ccfddl: cannot parse {path.name}: {exc}")
            continue
        if isinstance(loaded, dict):
            loaded = [loaded]
        for raw in loaded or []:
            if not isinstance(raw, dict):
                continue
            conference = _conference_of(raw)
            if conference is not None:
                out.append(conference)
    return out


class CcfddlSource:
    name = NAME

    def load(
        self, cache_dir: Path, *, offline: bool = False
    ) -> list[Conference]:
        root = fetch_tarball(REPO, REF, cache_dir, offline=offline)
        return parse_tree(root / "conference")

File: src/sources/ccfddl.ts
/**
 * ccfddl/ccf-deadlines source.
 * Ported from scripts/sources/ccfddl.py.
 */

import { readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { load as loadYaml } from "js-yaml";
import {
  type Conference,
  type Deadline,
  type DeadlineKind,
  type Edition,
  parseDateRange,
  parseInstant,
  slug,
  warn,
} from "../model.ts";
import { fetchTarball } from "./base.ts";

export const REPO = "ccfddl/ccf-deadlines";
export const REF = "main";
export const NAME = "ccfddl";

// 'abstract deadline' (with a space) exists once upstream.
const ABSTRACT_KEYS = ["abstract_deadline", "abstract deadline"];

function deadlinesOf(timeline: unknown, tzRaw: string): Deadline[] {
  const out: Deadline[] = [];
  for (const [index, entry] of ((timeline as unknown[] | null) ?? []).entries()) {
    if (typeof entry !== "object" || entry === null) continue;
    const rec = entry as Record<string, unknown>;
    const rnd = index + 1;
    const comment = rec.comment === null || rec.comment === undefined ? null : String(rec.comment);
    let rawAbstract: unknown = null;
    for (const key of ABSTRACT_KEYS) {
      if (rec[key] !== null && rec[key] !== undefined) {
        rawAbstract = rec[key];
        break;
      }
    }
    const candidates: Array<[DeadlineKind, string, unknown]> = [
      ["abstract", "Abstract submission", rawAbstract],
      ["paper", "Paper submission", rec.deadline],
    ];
    for (const [kind, label, raw] of candidates) {
      if (raw === null || raw === undefined) continue;
      const at = parseInstant(raw, tzRaw);
      if (at === null) continue;
      out.push({ kind, label, at_utc: at, tz_raw: tzRaw, round: rnd, comment });
    }
  }
  return out;
}

function editionOf(raw: Record<string, unknown>): Edition | null {
  const year = Number(raw.year);
  if (!Number.isInteger(year) || year <= 0) {
    warn(`ccfddl edition without a usable year: ${JSON.stringify(raw.id)}`);
    return null;
  }
  const tzRaw = String(raw.timezone ?? "");
  const dateText = String(raw.date ?? "");
  const [start, end] = parseDateRange(dateText, year);
  return {
    year,
    edition_id: String(raw.id ?? ""),
    link: String(raw.link ?? ""),
    place: String(raw.place ?? ""),
    date_text: dateText,
    event_start: start,
    event_end: end,
    deadlines: deadlinesOf(raw.timeline, tzRaw),
    estimated: false,
    source: NAME,
  };
}

function conferenceOf(raw: Record<string, unknown>): Conference | null {
  const title = String(raw.title ?? "").trim();
  if (!title) return null;
  const editions = ((raw.confs as unknown[] | null) ?? [])
    .map((c) =>
      typeof c === "object" && c !== null ? editionOf(c as Record<string, unknown>) : null,
    )
    .filter((e): e is Edition => e !== null)
    .sort((a, b) => a.year - b.year);
  const rank: Record<string, string> = {};
  for (const [k, v] of Object.entries((raw.rank as Record<string, unknown> | null) ?? {})) {
    if (v !== null && v !== undefined) rank[String(k).toLowerCase()] = String(v);
  }
  let link = "";
  for (const edition of [...editions].reverse()) {
    if (edition.link) {
      link = edition.link;
      break;
    }
  }
  return {
    key: slug(title),
    title,
    full_name: String(raw.description ?? title),
    link,
    rank,
    dblp: raw.dblp === null || raw.dblp === undefined ? null : String(raw.dblp),
    upstream_sub: raw.sub === null || raw.sub === undefined ? null : String(raw.sub),
    tags: [],
    categories: [],
    editions,
    sources: [NAME],
  };
}

/** Read every `conference/<SUB>/<name>.yml` under an extracted tree. */
export function parseTree(conferenceDir: string): Conference[] {
  const out: Conference[] = [];
  const files: string[] = [];
  const stack = [conferenceDir];
  while (stack.length > 0) {
    const cur = stack.pop();
    if (cur === undefined) break;
    for (const name of readdirSync(cur)) {
      const p = join(cur, name);
      if (statSync(p).isDirectory()) {
        stack.push(p);
      } else if (name.endsWith(".yml") && name !== "types.yml") {
        files.push(p);
      }
    }
  }
  for (const path of files.sort()) {
    let loaded: unknown;
    try {
      loaded = loadYaml(readFileSync(path, "utf8"));
    } catch (exc) {
      warn(`ccfddl: cannot parse ${path}: ${String(exc)}`);
      continue;
    }
    const items = Array.isArray(loaded) ? loaded : [loaded];
    for (const item of items) {
      if (typeof item !== "object" || item === null) continue;
      const conference = conferenceOf(item as Record<string, unknown>);
      if (conference !== null) out.push(conference);
    }
  }
  return out;
}

export class CcfddlSource {
  name = NAME;

  async load(cacheDir: string, options: { offline?: boolean } = {}): Promise<Conference[]> {
    const root = await fetchTarball(REPO, REF, cacheDir, options);
    return parseTree(join(root, "conference"));
  }
}
